File: sesion4-lab/aspnet-csrf/csrf-demo/Services/WalletService.cs
using System.Globalization;

namespace CsrfRazorDemo.Services;

/// <summary>
/// Saldo simulado en sesion HTTP (como una sesion bancaria autenticada).
/// </summary>
public class WalletService
{
    private const double InitialBalance = 1000.0;
    private const string BalanceKey = "wallet.balance";
    private const string MessageKey = "wallet.message";

    private readonly IHttpContextAccessor httpContextAccessor;

    public WalletService(IHttpContextAccessor httpContextAccessor)
    {
        this.httpContextAccessor = httpContextAccessor;
    }

    private ISession Session => httpContextAccessor.HttpContext!.Session;

    public double GetBalance()
    {
        var raw = Session.GetString(BalanceKey);
        return double.TryParse(raw, NumberStyles.Any, CultureInfo.InvariantCulture, out var balance)
            ? balance
            : InitialBalance;
    }

    public string GetLastTransferMessage() => Session.GetString(MessageKey) ?? "";

    public void Transfer(string to, double amount)
    {
        var balance = GetBalance();
        if (amount <= 0)
        {
            Session.SetString(MessageKey, "Importe invalido.");
            return;
        }
        if (amount > balance)
        {
            Session.SetString(MessageKey, "Saldo insuficiente.");
            return;
        }

        balance -= amount;
        Session.SetString(BalanceKey, balance.ToString(CultureInfo.InvariantCulture));
        Session.SetString(MessageKey, $"Transferidos {amount} EUR a {to}.");
    }

    public void Reset()
    {
        Session.SetString(BalanceKey, InitialBalance.ToString(CultureInfo.InvariantCulture));
        Session.Remove(MessageKey);
    }
}

File: sesion4-lab/spring-csrf/csrf-server/src/main/java/com/example/csrf/service/WalletService.java
package com.example.csrf.service;

import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.SessionScope;

/**
 * Saldo simulado en sesion HTTP (como una sesion bancaria autenticada).
 */
@Service
@SessionScope
public class WalletService {

  private static final double INITIAL_BALANCE = 1000.0;

  private double balance = INITIAL_BALANCE;
  private String lastTransferMessage = "";

  public double getBalance() {
    return balance;
  }

  public String getLastTransferMessage() {
    return lastTransferMessage;
  }

  public void transfer(String to, double amount) {
    if (amount <= 0) {
      lastTransferMessage = "Importe invalido.";
      return;
    }
    if (amount > balance) {
      lastTransferMessage = "Saldo insuficiente.";
      return;
    }
    balance -= amount;
    lastTransferMessage = "Transferidos " + amount + " EUR a " + to + ".";
  }

  public void reset() {
    balance = INITIAL_BALANCE;
    lastTransferMessage = "";
  }
}
